doc/architectural/background-concepts.md

```diff
@@ -1,32 +1,810 @@
+is executed several times, so it cannot be part of the first basic
+block. It cannot be part of the other basic blocks
+loop body, there is a branch right after it (meaning that no other
+statement is guaranteed to run right after it). The loop body clearly
+forms a basic block, as does the return instruction. 
+
+Putting everything together, we get a control flow graph like this:
+
+\dot "Control flow graph for factorial"
+digraph cfg {
+  node [shape=rectangle]
+  bb1 [label="Declaration: unsigned long fac.1, value 1\nDeclaration: unsigned i.1, value 1",peripheries=2]
+  bb2 [label="Expression: i.1 <= n.1"]
+  bb3 [label="Expression: fac.1 *= i.1\nExpression: i.1 ++",junk="*"]
+  
+  bb4 [label="Return: fac.1",style=filled,fillcolor=gray80]
+  {bb1, bb3} -> bb2
+  bb2 -> bb3 [label="true"]
+  bb2 -> bb4 [label="false"]
+}
+\enddot
+
+
+The graph can be read as follows: Each node corresponds to a basic
+block. The initial basic block (where the function is entered) is marked
+with a double border, while those basic blocks that leave the function
+have a gray background. The edges state that once execution arrives at
+the end of a basic block, it will follow one of the outgoing edges to
+another block. Some edges are labeled: The edge leaving the comparison
+basic block with the label `true`, for instance, can only be taken if
+the comparison did, in fact, return `true`.
+
+Note that this representation makes it very easy to interpret the
+program, keeping just two pieces of state: The current position (which
+basic block and which line), and the values of all variables (in real
+software, this would also include parts of the heap and the call stack).
+Execution proceeds as follows: As long as there are still instructions
+to execute in the current basic block, run the next instruction and move
+the current position to right after that instruction. At the end of a
+basic block, take one of the available outgoing edges to another basic block.
+This makes this representation a prime candidate to implement a number
+of static analyses, as described in section TODO.
 
 \subsection SSA_section SSA
 
-To be documented.
+While control flow graphs are already quite useful for static analysis,
+some techniques benefit from a further transformation to a
+representation know as **static single assignment**, short **SSA**. The
+point of this step is to ensure that we can talk about the entire history
+of assignments to a given variable. This is achieved by renaming
+variables again: Whenever we assign to a variable, we *clone* this
+variable by giving it a new name. This ensures that each variable
+appearing in the resulting program is written to exactly once. In this
+way, we can refer to earlier values of the variable by just referencing
+the name of an older incarnation of the variable.
+
+We illustrate this transformation by first showing how the body of the
+`for` loop of `factorial` is transformed. We currently have:
+
+```
+expression: fac.1 *= i.1
```

Comment:
Maybe use the appropriate @#$ suffixes used by ssa_exprt

Reply:
Where can I find documentation on this?

Reply:
Ooh, I remember making a PR to add the documentation. https://github.com/diffblue/cbmc/pull/886/files

Reply:
Hm, this carries quite a bit of additional information - should we really introduce all of this here? Especially since this seems a bit like an implementation detail.